File: lc-dc-sep/lc-dc-sep8/solution.cpp
// Problem: 725. Split Linked List in Parts

class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        vector<ListNode*> res;
        int n = 0;
        ListNode* curr = head;
        while(curr != nullptr){
            curr = curr->next;
            n++;
        }
        int splits = n / k;
        int rem = n % k;
        curr = head;
        for(int i=0; i<k; i++){
            ListNode newSplit(0);
            ListNode* tail = &newSplit;
            int currSplits = splits;
            if(rem > 0){
                rem--;
                currSplits++;
            }
            for(int j=0; j<currSplits; j++){
                tail->next = new ListNode(curr->val);
                tail = tail->next;
                curr = curr->next;
            }
            res.push_back(newSplit.next);
        }
        return res;
    }
};

File: lc-dc-sep/lc-dc-sep8/solution.py
# Problem: 725. Split Linked List in Parts

class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        res = []
        n = 0
        curr = head
        while(curr):
            curr = curr.next
            n += 1
        splits = n // k
        rem = n % k
        curr = head
        for i in range(k):
            newSplit = ListNode(0)
            tail = newSplit
            currSplits = splits
            if rem > 0:
                rem -= 1
                currSplits += 1
            for j in range(currSplits):
                tail.next = ListNode(curr.val)
                tail = tail.next
                curr = curr.next
            res.append(newSplit.next)

        return res
        